lib_c/dates_calc.c

```diff
@@ -0,0 +1,340 @@
+     arguments are [const].
+   - functions that can fail return the [dc_success] type. What is stored into
+     [ret] is unspecified when that is [dc_error].
+   - it is expected that [ret] and other arguments may overlap, so, in the code
+     below, a field in [ret] should never be written before the same field in
+     any other argument of the same type is read.
+*/
+
+typedef enum dc_success {
+  dc_error, dc_ok
+} dc_success;
+
+typedef enum dc_date_rounding {
+  dc_date_round_up,
+  dc_date_round_down,
+  dc_date_round_abort
+} dc_date_rounding;
+
+
+typedef struct dc_date {
+  long int year;
+  unsigned long int month;
+  unsigned long int day;
+} dc_date;
+
+typedef struct dc_period {
+  long int years;
+  long int months;
+  long int days;
+} dc_period;
+
+void dc_make_period(dc_period *ret, const long int y, const long int m, const long int d) {
+  ret->years = y;
+  ret->months = m;
+  ret->days = d;
+}
+
+void dc_print_period (const dc_period *p) {
+  printf("[%ld years, %ld months, %ld days]", p->years, p->months, p->days);
+}
+
+void dc_period_of_string (dc_period *ret, const char* s) {
+  sscanf("[%ld years, %ld months, %ld days]", s, &ret->years, &ret->months, &ret->days);
+}
+
+void dc_add_periods (dc_period *ret, const dc_period *p1, const dc_period *p2) {
+  ret->years = p1->years + p2->years;
+  ret->months = p1->months + p2->months;
+  ret->days = p1->days + p2->days;
+}
+
+void dc_sub_periods (dc_period *ret, const dc_period *p1, const dc_period *p2) {
+  ret->years = p1->years - p2->years;
+  ret->months = p1->months - p2->months;
+  ret->days = p1->days - p2->days;
+}
+
+void dc_mul_periods (dc_period *ret, const dc_period *p, const long int m) {
+  ret->years = p->years * m;
+  ret->months = p->months * m;
+  ret->days = p->days * m;
+}
+
+dc_success dc_period_to_days (long int *ret, const dc_period *p) {
+  if (p->years || p->months) return dc_error;
+  else *ret = p->days;
+  return dc_ok;
+}
+
+BOOL dc_is_leap_year (const long int y) {
+  return (y % 400 == 0 || (y % 4 == 0 && y % 100 != 0));
+}
+
+unsigned long int dc_days_in_month (const dc_date *d) {
+  switch (d->month) {
+  case 2:
+    return (dc_is_leap_year(d->year) ? 29 : 28);
+  case 4: case 6: case 9: case 11:
+    return 30;
+  default:
+    return 31;
+  }
+}
+
+BOOL dc_is_valid_date (const dc_date *d) {
+  return (1 <= d->day && d->day <= dc_days_in_month(d));
+}
+
+dc_success dc_make_date(dc_date *ret, const long int y, const unsigned long int m, const unsigned long int d) {
+  ret->year = y;
+  ret->month = m;
+  ret->day = d;
+  if (dc_is_valid_date(ret)) return dc_ok;
+  else return dc_error;
+}
+
+void dc_copy_date(dc_date *ret, const dc_date *d) {
+  if (ret != d) {
+    ret->year = d->year;
+    ret->month = d->month;
+    ret->day = d->day;
+  }
+}
+
+/* Precondition: [1 <= d->month <= 12]. The returned day is always [1] */
+void dc_add_months(dc_date *ret, const dc_date *d, const long int months) {
```

Comment:
In OCaml `add_months_to_first_of_month_date` is recursive, I think it is missing here? (ie when `|months| >= 12`)

Reply:
hm, indeed, I changed how this function works a bit and that strays away from what's been proved.
This implementation uses a direct arithmetic computation that *should* be equivalent to the recursive definition.
And it was actually wrong because it used modulo computation on months in range [1, 12] ; hopefully just shifting by -1 during the computation should ensure a valid date is returned.